test/sslapitest.c

```diff
@@ -1164,85 +1157,409 @@ static int test_ktls_sendfile(void)
+    case 4:
+       return test_ktls_no_txrx_client_no_rx_server(TLS1_2_VERSION);
+    case 5:
+       return test_ktls_no_rx_client_no_rx_server(TLS1_2_VERSION);
+    case 6:
+       return test_ktls_no_tx_client_no_rx_server(TLS1_2_VERSION);
+    case 7:
+       return test_ktls_client_no_rx_server(TLS1_2_VERSION);
+    case 8:
+       return test_ktls_no_txrx_client_no_tx_server(TLS1_2_VERSION);
+    case 9:
+       return test_ktls_no_rx_client_no_tx_server(TLS1_2_VERSION);
+    case 10:
+       return test_ktls_no_tx_client_no_tx_server(TLS1_2_VERSION);
+    case 11:
+       return test_ktls_client_no_tx_server(TLS1_2_VERSION);
+    case 12:
+       return test_ktls_no_txrx_client_server(TLS1_2_VERSION);
+    case 13:
+       return test_ktls_no_rx_client_server(TLS1_2_VERSION);
+    case 14:
+       return test_ktls_no_tx_client_server(TLS1_2_VERSION);
+    case 15:
+       return test_ktls_client_server(TLS1_2_VERSION);
+    default:
+       return 0;
+    }
 }
 
-static int test_ktls_no_rx_client_server(void)
+static int test_ktls_sendfile_tls12(int tst)
 {
-    return execute_test_ktls(1, 0, 1, 1);
+   char *cipher[] = {"AES128-GCM-SHA256","AES128-CCM-SHA256","AES256-GCM-SHA384"};
+#ifndef OPENSSL_KTLS_AES_GCM_128
+   if(tst == 1) return 1;
+#endif
+#ifndef OPENSSL_KTLS_AES_CCM_128
+   if(tst == 2) return 1;
+#endif
+#ifndef OPENSSL_KTLS_AES_GCM_256
+   if(tst == 3) return 1;
+#endif
+   return test_ktls_sendfile(TLS1_2_VERSION,cipher[tst]);
 }
 
-static int test_ktls_no_tx_client_server(void)
+#endif
+
+#if !defined(OPENSSL_NO_TLS1_3)
+static int test_ktls_tls13(int test)
 {
-    return execute_test_ktls(0, 1, 1, 1);
+    switch(test) {
+    case 0:
+        return test_ktls_no_txrx_client_no_txrx_server(TLS1_3_VERSION);
+    case 1:
+       return test_ktls_no_rx_client_no_txrx_server(TLS1_3_VERSION);
+    case 2:
+       return test_ktls_no_tx_client_no_txrx_server(TLS1_3_VERSION);
+    case 3:
+       return test_ktls_client_no_txrx_server(TLS1_3_VERSION);
+    case 4:
+       return test_ktls_no_txrx_client_no_rx_server(TLS1_3_VERSION);
+    case 5:
+       return test_ktls_no_rx_client_no_rx_server(TLS1_3_VERSION);
+    case 6:
+       return test_ktls_no_tx_client_no_rx_server(TLS1_3_VERSION);
+    case 7:
+       return test_ktls_client_no_rx_server(TLS1_3_VERSION);
+    case 8:
+       return test_ktls_no_txrx_client_no_tx_server(TLS1_3_VERSION);
+    case 9:
+       return test_ktls_no_rx_client_no_tx_server(TLS1_3_VERSION);
+    case 10:
+       return test_ktls_no_tx_client_no_tx_server(TLS1_3_VERSION);
+    case 11:
+       return test_ktls_client_no_tx_server(TLS1_3_VERSION);
+    case 12:
+       return test_ktls_no_txrx_client_server(TLS1_3_VERSION);
+    case 13:
+       return test_ktls_no_rx_client_server(TLS1_3_VERSION);
+    case 14:
+       return test_ktls_no_tx_client_server(TLS1_3_VERSION);
+    case 15:
+       return test_ktls_client_server(TLS1_3_VERSION);
```

Comment:
test_ktls_tls13 seems to be identical to test_ktls_tls12 except for the parameter sent to all of the functions. Surely there is come opportunity to rationalise this by merging the two functions.

Reply:
Refactored. The same change done to test_ktls_sendfile_tls{12,13} functions.